homeassistant/components/vacuum/roomba.py

```diff
@@ -0,0 +1,301 @@
+
+SUPPORT_ROOMBA = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_PAUSE | \
+                 SUPPORT_STOP | SUPPORT_RETURN_HOME | SUPPORT_BATTERY | \
+                 SUPPORT_STATUS | SUPPORT_SEND_COMMAND
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the iRobot Roomba vacuum cleaner platform."""
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
+
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    certificate = config.get(CONF_CERT)
+    continuous = config.get(CONF_CONTINUOUS)
+
+    # Create handler
+    roomba = RoombaVacuum(
+        hass, name, host, username, password, certificate, continuous)
+    hass.data[PLATFORM][host] = roomba
+
+    async_add_devices([roomba], update_before_add=True)
+
+
+class RoombaVacuum(VacuumDevice):
+    """Representation of a Xiaomi Vacuum cleaner robot."""
+
+    def __init__(self, hass, name, host, username, password, certificate,
+                 continuous):
+        """Initialize the Roomba handler."""
+        self.hass = hass
+        self._name = name
+        self._icon = ICON
+        self._host = host
+        self._username = username
+        self._password = password
+        self._certificate = certificate
+        self._continuous = continuous
+        self._vacuum = None
+        self._battery_level = None
+        self._status = None
+        self._state_attrs = {}
+        self.vacuum_state = None
+        self._is_on = False
+        self._available = False
+        self._metric = hass.config.units.is_metric
+
+    @property
+    def supported_features(self):
+        """Flag vacuum cleaner robot features that are supported."""
+        return SUPPORT_ROOMBA
+
+    @property
+    def battery_level(self):
+        """Return the battery level of the vacuum cleaner."""
+        return self._battery_level
+
+    @property
+    def status(self):
+        """Return the status of the vacuum cleaner."""
+        return self._status
+
+    @property
+    def state(self) -> str:
+        """Return the state."""
+        return STATE_ON if self.is_on else STATE_OFF
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self._is_on
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use for device."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._state_attrs
+
+    @property
+    def vacuum(self):
+        """Property accessor for vacuum object."""
+        if not self._vacuum:
```

Comment:
This way of accessing the vacuum handler was discouraged in #8623, so, if you're getting it only one time, do it at setup, and look for the I/O operations (`self._vacuum.connect()`?) to be async friendly